packages/sheets-ui/src/controllers/auto-width.controller.ts

```diff
@@ -0,0 +1,108 @@
+
+export const AFFECT_LAYOUT_STYLES = ['ff', 'fs', 'tr', 'tb'];
+
+export class AutoWidthController extends Disposable {
+    constructor(
+        @IRenderManagerService private readonly _renderManagerService: RenderManagerService,
+        @Inject(SheetInterceptorService) private readonly _sheetInterceptorService: SheetInterceptorService,
+        @Inject(IUniverInstanceService) private readonly _univerInstanceService: IUniverInstanceService
+    ) {
+        super();
+        this._initialize();
+    }
+
+    getUndoRedoParamsOfColWidth(ranges: IRange[]) {
+        const { _univerInstanceService: univerInstanceService } = this;
+
+        const workbook = univerInstanceService.getCurrentUnitForType<Workbook>(UniverInstanceType.UNIVER_SHEET)!;
+        const worksheet = workbook.getActiveSheet();
+        const unitId = workbook.getUnitId();
+        const subUnitId = worksheet.getSheetId();
+
+        const sheetSkeletonService = this._renderManagerService.getRenderById(unitId)!.with<SheetSkeletonManagerService>(SheetSkeletonManagerService);
+        if (!subUnitId || !sheetSkeletonService.getCurrent()) {
+            return {
+                redos: [],
+                undos: [],
+            };
+        }
+        const { skeleton } = sheetSkeletonService.getCurrent()!;
+        const colsAutoWidthInfo: IColAutoWidthInfo[] = skeleton.calculateAutoWidthInRange(ranges);
+
+        const colWidthObject: IObjectArrayPrimitiveType<Nullable<number>> = {};
+        for (const { col, width } of colsAutoWidthInfo) {
+            colWidthObject[col] = width;
+        }
+        const redoParams: ISetWorksheetColWidthMutationParams = {
+            subUnitId,
+            unitId,
+            ranges,
+            colWidth: colWidthObject,
+        };
+        const undoParams: ISetWorksheetColWidthMutationParams = SetWorksheetColAutoWidthMutationFactory(redoParams, worksheet);
+        return {
+            undos: [
+                {
+                    id: SetWorksheetColWidthMutation.id,
+                    params: undoParams,
+                },
+            ],
+            redos: [
+                {
+                    id: SetWorksheetColWidthMutation.id,
+                    params: redoParams,
+                },
+            ],
+        };
+    }
+
+    private _initialize() {
+        const { _sheetInterceptorService: sheetInterceptorService } =
+            this;
+
+        // intercept 'sheet.command.set-col-is-auto-width' command.
+        this.disposeWithMe(sheetInterceptorService.interceptCommand({
+            getMutations: (command: { id: string; params: ISetWorksheetColIsAutoWidthMutationParams }) => {
+                if (command.id !== SetWorksheetColIsAutoWidthCommand.id) {
+                    return {
+                        redos: [],
+                        undos: [],
+                    };
+                }
+
+                return this.getUndoRedoParamsOfColWidth(command.params.ranges);
```

Comment:
Should also add `unitId` and `subUnitId` as params. We shouldn't rely on the UI state if we have specific params.

Reply:
<img width="784" alt="image" src="https://github.com/user-attachments/assets/907f70e1-d185-4b04-949a-3dc83cf975e7">
 Yes, untId & subUnitId is in return value